### PR TITLE
chore!: develop on Node 20, CI with 16.17.1 + 18.17.1 + 20

### PR DESCRIPTION
**Pull request by EvanHahn**
_I recommend [reviewing this with whitespace changes disabled](https://github.com/digidem/mapeo-map-server/pull/103/files?w=1)._

This change updates this project's Node versions:

- Develop on Node 20 by updating `.nvmrc`.

- Require Node 16.17.1 in the `engines` field of `package.json`.

- CI should test on Node 20, 18.17.1, and 16.17.1.

There are four other changes in this upgrade:

1. `package-lock.json` was updated because of a new npm version.

2. better-sqlite3 wouldn't compile on Node 20 so I updated it to a newer version.

3. Starting a test server caused IPv6 issues, so I explicitly used IPv4.

4. Two of our `EventSource` tests did not work. I spent a lot of time trying to figure out why, but I suspect it's a bug in the `eventsource` package because it worked with [Undici's version][1]. Unfortunately, Undici requires Node 18+ so we can't use it. After several hours, I replaced the tests with simpler HTTP requests and things passed.

[0]: https://github.com/nodejs-mobile/nodejs-mobile-react-native/tree/51afe46d36d3776d1bfe7158073897216b55d7f2#versioning
[1]: https://undici.nodejs.org/#/docs/api/EventSource

### CODE REVIEW

File: test/e2e/imports.test.js [hunk 1]
```diff
@@ -177,59 +147,6 @@ test('GET /imports/progress/:importId when import is already completed returns s
   t.same(messages2[0], messages1[messages1.length - 1])
 })
 
-// This tests that the server can force an eventsource client to disconnect by
-// responding with a 204 status code. This is in case the client does not close
-// the eventSource (as it should) after the 'complete' message is received
-test('GET /imports/progress/:importId - EventSource forced to close after single message if import has already completed', async (t) => {
```
Comment (EvanHahn) [inline test/e2e/imports.test.js:183]: This test seemed nearly identical to the one above _and_ it wasn't working, so I removed it entirely.

**Comment by achou11** [comment]
Is this PR completely removing the event source tests? slightly concerned about doing that because the design of the import progress endpoint is be used by an event source approach from a client

File: package.json [hunk 2]
```diff
@@ -56,7 +56,7 @@
     "uuid": "^8.3.2"
   },
   "peerDependencies": {
-    "better-sqlite3": ">=7.6.0"
+    "better-sqlite3": ">=9.3.0"
```
Comment (achou11) [inline package.json:59]: generally okay with this change but we have to keep in mind that if we incorporate this as part of the mobile app, we want to be able to dedupe node-deps containing native pieces as much as possible with core next. given that, we usually had this set to something that was compatible with core (or else we'd have to make sure the version used in core is compatible with this one)

**Comment by EvanHahn** [comment]
No, it keeps most of the EventSource tests.

We had two tests which (1) broke after the Node upgrade (2) were nearly identical. These test that the server returns a 204 if the `Last-Event-ID` is "complete" or "error". I removed one of the tests and replaced the other with a "regular" HTTP request rather than an EventSource.

Happy to attempt something different (though I spent a few hours just trying to get these tests to work and am not sure of an alternative; would love to pair on it if so).

File: test/test-helpers/create-server.js [hunk 3]
```diff
@@ -18,16 +18,20 @@ module.exports = createServer
...
-  const server = createMapServer(
-    { logger: false, forceCloseConnections: true },
-    { database: new Db(dbPath) }
+  const server = Object.assign(
+    createMapServer(
+      { logger: false, forceCloseConnections: true },
+      { database: new Db(dbPath) }
+    ),
+    // Use 127.0.0.1 to avoid IPv6 connection issues.
+    { testListen: () => server.listen(0, '127.0.0.1') }
```
Comment (achou11) [inline test/test-helpers/create-server.js:34]: is this needed? think you can specify the host using the fastify instance's  `listen()` method: https://fastify.dev/docs/v3.29.x/Reference/Server/#listen

File: test/e2e/imports.test.js (hunk 1 above)
Comment (achou11) [reply test/e2e/imports.test.js:183]: huh weird - but yeah, seems very similar 😄

**Review by achou11 — CHANGES_REQUESTED** [review]
Surprised that this is all that's needed for this! Would note somewhere in the commit title that this is a breaking change

File: package.json (hunk 2 above)
Comment (EvanHahn) [reply package.json]: Good thought. Downgraded in 60cca7536a252dda0d1361d6921c4c4b911a38d4.

File: test/test-helpers/create-server.js (hunk 3 above)
Comment (EvanHahn) [reply test/test-helpers/create-server.js]: Done in 30975a367075eaab1e682d0737a125bc7005855e.

**Review by achou11 — APPROVED** [review]

**Pull request merged**